packages/router/src/scope.ts

```diff
@@ -186,60 +189,30 @@ export class Scope {
     };
   }
 
-  // public findViewport(name: string): Viewport {
-  //   const parts = name.split(this.router.separators.scope);
-  //   const names = parts.shift().split(this.router.separators.viewport);
-  //   const comp = names.shift();
-  //   name = names.shift();
-  //   let newScope = false;
-  //   if (name.endsWith(this.router.separators.ownsScope)) {
-  //     newScope = true;
-  //     name = name.substring(0, name.length - 1);
-  //   }
-  //   const viewport = this.resolveViewport(name, comp) || this.addViewport(name, null, { scope: newScope });
-  //   if (!parts.length) {
-  //     return viewport;
-  //   } else {
-  //     const scope = viewport.scope || viewport.owningScope;
-  //     return scope.findViewport(parts.join(this.router.separators.scope));
-  //   }
-  // }
-
-  // public resolveViewport(name: string, component: string): Viewport {
-  //   if (name.length && name.charAt(0) !== '?') {
-  //     return this.viewports[name];
-  //   }
-  //   // Need more ways to resolve viewport based on component name!
-  //   const comp = this.resolveComponent(component);
-  //   if (comp.viewport) {
-  //     name = comp.viewport;
-  //     return this.viewports[name];
-  //   }
-  //   return null;
-  // }
-
-  public addViewport(name: string, element: Element, options?: IViewportOptions): Viewport {
+  public addViewport(name: string, element: Element, container: IRenderContext, options?: IViewportOptions): Viewport {
     let viewport = this.viewports[name];
     if (!viewport) {
       let scope: Scope;
       if (options.scope) {
-        scope = new Scope(this.router, element, this);
+        scope = new Scope(this.router, element, container, this);
         this.router.scopes.push(scope);
       }
 
-      viewport = this.viewports[name] = new Viewport(this.router, name, element, this, scope, options);
+      viewport = this.viewports[name] = new Viewport(this.router, name, element, container, this, scope, options);
     }
-    if (element) {
-      viewport.setElement(element, options);
+    if (element || container) {
+      viewport.setElement(element, container, options);
     }
     return viewport;
   }
-  public removeViewport(viewport: Viewport): number {
-    if (viewport.scope) {
-      this.router.removeScope(viewport.scope);
+  public removeViewport(viewport: Viewport, element: Element, container: IRenderContext): number {
+    if ((!element && !container) || viewport.remove(element, container)) {
+      if (viewport.scope) {
+        this.router.removeScope(viewport.scope);
+      }
+      // tslint:disable-next-line:no-dynamic-delete
+      delete this.viewports[viewport.name];
```

Comment:
```suggestion
      Reflect.deleteProperty(this.viewports, viewport.name);
```
Then you can remove the suppression.

Reply:
Not sure that's my preference, but on the other hand I don't feel that much about it. Fixed all the other deletes as well.

Reply:
It takes some getting used to, I didn't particularly like it either, but I'm starting to see its elegance. It is kind of the way forward with the latest ES which we want to try to embrace.